bundles/config/org.eclipse.smarthome.config.xml/src/main/java/org/eclipse/smarthome/config/xml/AbstractXmlBasedProvider.java

```diff
@@ -0,0 +1,231 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.config.xml;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.osgi.framework.Bundle;
+
+/**
+ * Common base class for XML based providers.
+ *
+ * @author Simon Kaufmann - initial contribution, factored out of subclasses
+ *
+ * @param <T_ID> the key type, e.g. ThingTypeUID, ChannelUID, URI,...
+ * @param <T_OBJECT> the object type, e.g. ThingType, ChannelType, ConfigDescription,...
+ */
+public abstract class AbstractXmlBasedProvider<T_ID, T_OBJECT> {
+
+    private static class LocalizedKey {
+        public final Object id;
+        public final String locale;
+
+        public LocalizedKey(Object id, String locale) {
+            this.id = id;
+            this.locale = locale;
+        }
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + ((id == null) ? 0 : id.hashCode());
+            result = prime * result + ((locale == null) ? 0 : locale.hashCode());
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            LocalizedKey other = (LocalizedKey) obj;
+            if (!Objects.equals(id, other.id)) {
+                return false;
+            }
+            if (!Objects.equals(locale, other.locale)) {
+                return false;
+            }
+            return true;
+        }
+
+    }
+
+    private final Map<Bundle, List<T_OBJECT>> bundleObjectMap = new ConcurrentHashMap<>();
+    private final Map<LocalizedKey, T_OBJECT> localizedObjectCache = new ConcurrentHashMap<>();
+
+    /**
+     * Return the unique identifier of the given object (e.g. UID, URI,...)
+     *
+     * @param object the object to return the identifier for
+     * @return the identifier. Must not be <code>null</code>
+     */
+    protected abstract T_ID getIndentifier(T_OBJECT object);
```

Comment:
this will probably be replaced by getUID from `Identifiable`, but if not, you might want to fix the typo (Indent vs. Ident).